### PR TITLE
automatic support for host-to-host networking that only does link-local addressing

### PR DESCRIPTION
**Pull request by poettering**
This adds some basic support for automatic host-to-host networking (i.e. direct connections between two systems with network connectivity that does not involve the Internet).

This is a generalization of https://github.com/NetworkManager/NetworkManager/blob/main/data/90-nm-thunderbolt.rules#L11 but also covers Prolific host-to-host USB devices.

### CODE REVIEW

**Comment by poettering** [comment]
(I also notified NM folks about this: https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/issues/1413)

File: rules.d/82-net-auto-link-local.rules [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# do not edit this file, it will be overwritten on update
+
+ACTION=="remove", GOTO="net_link_local_end"
+SUBSYSTEM!="net", GOTO="net_link_local_end"
+
+# Network interfaces for which only Link-Local communication (i.e. IPv4LL, …)
+# makes sense, because they almost certainy will point to another host, not an
+# internet router.
+
+# (Note: matches against VID/PID go into 82-net-auto-link-local.hwb instead)
```
Comment (yuwata) [inline rules.d/82-net-auto-link-local.rules]: hwb -> hwdb

File: network/80-auto-link-local.network.example [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+# This example config file is installed as part of systemd.
+# It may be freely copied and edited (following the MIT No Attribution license).
+#
+# To use the file, one of the following methods may be used:
+# 1. add a symlink from /etc/systemd/network to the current location of this file,
+# 2. copy the file into /etc/systemd/network or one of the other paths checked
+#    by systemd-networkd and edit it there.
+# This file should not be edited in place, because it'll be overwritten on upgrades.
+
+# Enable IPV4LL (but no DHCP) for local links
```
Comment (yuwata) [inline network/80-auto-link-local.network.example]: IPV4LL -> IPv4LL

File: network/80-auto-link-local.network.example [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+[Match]
+Type=ether
+Property=ID_NET_AUTO_LINK_LOCAL_ONLY=1
+
+[Network]
+DHCP=yes
```
Comment (yuwata) [inline network/80-auto-link-local.network.example]: Hm?

**Review by yuwata — CHANGES_REQUESTED** [review]

File: network/80-auto-link-local.network.example (hunk 3 above)
Comment (poettering) [reply network/80-auto-link-local.network.example]: uh

**Comment by yuwata** [comment]
Please also disable IPv6AcceptRA=, which is enabled by default.

File: TODO [hunk 4]
```diff
@@ -133,16 +133,11 @@ Deprecations and removals:
...
-  particularly useful in storage target mode (i.e. NVME-TCP) modes. i.e. do
-  what https://christian.kellner.me/2018/05/24/thunderbolt-networking-on-linux/
-  and
-  https://github.com/NetworkManager/NetworkManager/blob/main/data/90-nm-thunderbolt.rules#L11
-  do for NM, but generically.
+* networkd: controlled by build-time options install some of the
+  *.network.example files we install currently as regular files *.network. For
+  many cases it's what you want: just configure the damn network. Hopefully
+  this can be enabled on Fedora even, though into a separate sub-RPM, so that
+  we don't conflict with NM.
```
Comment (thom311) [inline TODO:140]: I don't think there would be a conflict with NetworkManager.

The udev attribute has no effect to NetworkManager, unless it's explicitly honored. And then NetworkManager needs to honor it in a way that makes sense for that attribute.

Comment (poettering) [reply TODO:140]: @thom311 oh, well, this is mostly a comment on the 80-ethernet.network.example we ship, which just does DHCP on anything that looks like ethernet. If we'd install that unconditionally, then any wired ethernet interface would start being fought for by NM+networkd.

**Review by yuwata — APPROVED** [review]

**Pull request merged**